### PR TITLE
[expo-image] Add resizeMode support

### PR DESCRIPTION
**Pull request by IjzerenHein**
# Why

`expo-image` should resize according to the `resizeMode` prop. When no `resizeMode` is provided, the image uses `cover`.

# How

## JS
- `resizeMode` prop is removed from style

## iOS
- Setting `UIImageView.contentMode` adds support for `cover`, `contain` and `stretch`
- Setting the SDWebImage scale-factor was required for `center` and `repeat`
- Adding a custom load completion has was needed to set the "tile" mode on the UIImage, in order to make `repeat` work

## Android
- Setting the `ImageView.scaleType` and providing Glide with the option `fitCenter` adds support for `cover`, `contain` and `stretch`
- Overriding the image-size in the Glide options add support for `center`
- `repeat` is not yet supported

## Web
- No effort was needed

# Test Plan

- All images (also Animated gifs) now render with the same size as `<Image>` (tested with the test-app in NCL)
- Overflow issues are out of scope of this PR
- The only exception is `repeat` on Android with requires additional effort and is not supported yet

Simulator Screen Shot - iPhone 11 Pro Max - 2020-03-16 at 10 41 06
Screenshot_1584352026

### CODE REVIEW

File: packages/expo-image/ios/EXImageView.m [hunk 1]
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
...
 {
-  NSURL *imageURL = [RCTConvert NSURL:source[sourceUriKey]];
+  _source = source;
+  _needsReload = YES;
+}
 
-  [self sd_setImageWithURL:imageURL];
+- (void)setResizeMode:(RCTResizeMode)resizeMode
+{
+  if (_resizeMode == resizeMode) return;
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]: Let's write out full-sized `if` with brackets (💅)

File: packages/expo-image/ios/EXImageView.m [hunk 2]
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
...
+  if (_needsReload) {
+    [self reloadImage];
+  }
+}
+
+- (void)reloadImage
+{
+  _needsReload = NO;
+  
+  NSURL *imageUrl = _source ? [RCTConvert NSURL:_source[sourceUriKey]] : nil;
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]: At first I also used `imageUrl`, but then I noticed that both `NSURL` treats `URL` as full uppercase and there is a `sd_imageURL`, so I figured that must be the Obj-C way of naming URLs (💅)

File: packages/expo-image/ios/EXImageView.m [hunk 3]
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
...
+    
+    // Modifications to the image like changing the resizing-mode or cap-insets
+    // cannot be handled using a SDWebImage transformer, because they don't change
+    // the image-data and this causes this "meta" data to be lost in the SWWebImage caching process.
+    if (image) {
+      if (resizeMode == RCTResizeModeRepeat) {
+        image = [image resizableImageWithCapInsets:UIEdgeInsetsZero resizingMode:UIImageResizingModeTile];
+      }
+    }
+    self.image = image;
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]: Strongly catching `self` may leak its instance. Somehow we guard against that in `dealloc`, cancelling all operations, but still, we should use _weak self_ pattern.

File: packages/expo-image/ios/EXImageView.m [hunk 4]
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
...
+  [self sd_setImageWithURL:imageUrl
+          placeholderImage:nil
+                   options:SDWebImageAvoidAutoSetImage
+                   context:context
+                  progress:nil
+                 completed:^(UIImage * _Nullable image, NSError * _Nullable error, SDImageCacheType cacheType, NSURL * _Nullable imageURL) {
+    
+    // Modifications to the image like changing the resizing-mode or cap-insets
+    // cannot be handled using a SDWebImage transformer, because they don't change
+    // the image-data and this causes this "meta" data to be lost in the SWWebImage caching process.
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]:
```suggestion
    // the image-data and this causes this "meta" data to be lost in the SDWebImage caching process.
```

File: packages/expo-image/ios/EXImageView.m (hunk 3 above)
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]: Do we want to reset the `image` property even if it isn't changed when `resizeMode == repeat`? In general `sd_setImageWithURL` takes care of `self.image = image` and I may be afraid (having your comment in mind from the other PR) that setting `image` property may cause some extra operations on this `UIImageView`.

EDIT: Ok, I now saw the `SDWebImageAvoidAutoSetImage`. Do you think it would be worth to comment about it here? 😅

File: packages/expo-image/ios/EXImageView.m [hunk 5]
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
...
+{
+  if (_resizeMode == resizeMode) return;
+  
+  // Image needs to be reloaded whenever repeat is enabled or disabled
+  _needsReload = _needsReload || (resizeMode == RCTResizeModeRepeat) || (_resizeMode == RCTResizeModeRepeat);
+  _resizeMode = resizeMode;
+  
+  // Repeat resize mode is handled by the UIImage. Use scale to fill
+  // so the repeated image fills the UIImageView.
+  self.contentMode = resizeMode == RCTResizeModeRepeat
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]: Maybe we should move this change to the completion block of the `sd_setImageWithURL`?

File: packages/expo-image/ios/EXImageView.m [hunk 6]
```diff
@@ -17,9 +36,67 @@ - (void)dealloc
 
 - (void)setSource:(NSDictionary *)source
 {
-  NSURL *imageURL = [RCTConvert NSURL:source[sourceUriKey]];
+  _source = source;
+  _needsReload = YES;
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]:
```suggestion
  // TODO: Implement equatable image source abstraction
  _needsReload = YES;
```
? 😃

File: packages/expo-image/ios/EXImageView.m [hunk 7]
```diff
@@ -4,9 +4,28 @@
...
+@end
 
 @implementation EXImageView
 
+- (instancetype)init
+{
+  if (self = [super init]) {
+    _needsReload = NO;
+    _resizeMode = RCTResizeModeCover;
+    self.contentMode = (UIViewContentMode)_resizeMode;
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]: This may be just me, but I'm not particularly fond of just force casting that way. Do you think extracting this to a helper function `contentModeFromResizeMode` would be an overkill?

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageTypes.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+import android.widget.ImageView;
+import android.widget.ImageView.ScaleType;
+
+import com.facebook.react.bridge.JSApplicationIllegalArgumentException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class ExpoImageTypes {
```
Comment (sjchmiela) [inline packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageTypes.java]: When implementing `expo-notifications` I had to add several enum types to the package and I came up with this _enum pattern_, where for every JS-native enum we define an enum + a couple of helper functions. This has some advantages:
- an instance of an enum holds information about both JS and native value (coherency)
- putting different enums into different classes doesn't add too much bloat to one file
- with colocated functions we can convert back and forth between values of different realms
- you can use anything as native type (Strings too)
- an extra layer of abstraction between JS enum and native values (not exporting `ScaleType.FIT_CENTER` directly to JS)

Disadvantages of current implementations is `o(n)` searching when converting (could be mitigated with a simple hashmap creation).

I haven't tried to push to make this a convention in this repository, but so far I like it. What do you think about this pattern? An example would be:

https://github.com/expo/expo/blob/d09fc44c62ae96e0347e1909a774b1b93d866b49/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/enums/AudioContentType.java

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java [hunk 9]
```diff
@@ -40,9 +44,34 @@ public void setSource(ImageView view, @Nullable ReadableMap sourceMap) {
...
+
+  @ReactProp(name = "resizeMode")
+  public void setResizeMode(ImageView view, String resizeMode) {
+    ImageView.ScaleType scaleType;
+    if (resizeMode.equals("repeat")) {
+      scaleType = ImageView.ScaleType.FIT_XY;
+    } else {
+      scaleType = ExpoImageTypes.getScaleType(resizeMode);
+    }
+    view.setScaleType(scaleType);
```
Comment (sjchmiela) [inline packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java]: This implementation differs from the Android one when it comes to when the prop changes are applied. We could probably let it go, if this PR lands before the events PR I will simply rebase and split these changes.

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java [hunk 10]
```diff
@@ -40,9 +44,34 @@ public void setSource(ImageView view, @Nullable ReadableMap sourceMap) {
       return;
     }
 
+    RequestOptions options = new RequestOptions();
+
+    // Override the size for local assets. This ensures that
```
Comment (sjchmiela) [inline packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java]: How (does?) does it work for remote assets? 😕

**Review by sjchmiela — COMMENTED** [review]
Thanks so much for this work! The results are looking great! 👍

File: packages/expo-image/src/Image.tsx [hunk 11]
```diff
@@ -11,10 +18,16 @@ export interface ImageProps extends AccessibilityProps {
...
+  resizeMode?: ImageResizeMode;
 }
 
 export default class Image extends React.Component<ImageProps> {
   render() {
-    return <ExpoImage {...this.props} />;
+    const { style, resizeMode: resizeModeProp, ...restProps } = this.props;
+
+    const { resizeMode: resizeModeStyle, ...restStyle } = StyleSheet.flatten([style]) || {};
+    const resizeMode = resizeModeProp || resizeModeStyle || 'cover';
```
[the remark this answers is not in the document]
Comment (IjzerenHein) [reply packages/expo-image/src/Image.tsx]: Will do!

File: packages/expo-image/src/ExpoImage.tsx [hunk 12]
```diff
@@ -1,12 +1,18 @@
...
 
 const NativeExpoImage = requireNativeComponent('ExpoImage');
 
-export default function ExpoImage({ source, ...props }: ImageProps) {
+export default function ExpoImage({ source, style, ...props }: ImageProps) {
   const resolvedSource = Image.resolveAssetSource(source ?? {});
+  let resolvedStyle = style;
 
-  return <NativeExpoImage {...props} source={resolvedSource} />;
+  if (!Array.isArray(resolvedSource)) {
```
[the remark this answers is not in the document]
Comment (IjzerenHein) [reply packages/expo-image/src/ExpoImage.tsx]: Yes I believe so. The `source` as defined in the RN Image component can be an array of `ImageURISource`. These can be multiple sources and the native code can then select the best option for the given resolution.
https://reactnative.dev/docs/image/#source
https://github.com/facebook/react-native/blob/master/Libraries/Image/ImageSource.js#L85

The typeof of such an array is `object` and `resolveAssetSource` will just return that if that is the case.

File: packages/expo-image/src/ExpoImage.tsx [hunk 13]
```diff
@@ -1,12 +1,18 @@
...
+    const { width, height } = resolvedSource;
+    resolvedStyle = StyleSheet.flatten([{ width, height }, style]);
```
[the remark this answers is not in the document]
Comment (IjzerenHein) [reply packages/expo-image/src/ExpoImage.tsx]: This code was taken primarily from the RN source:
https://github.com/facebook/react-native/blob/master/Libraries/Image/Image.ios.js#L103

I believe that it is used for Yoga layout purposes only and it only applies to `require(..)` assets. These assets contain the width/height and form the default when laying out the image component.

File: packages/expo-image/ios/EXImageView.m (hunk 1 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: You're right, will do 👍

File: packages/expo-image/ios/EXImageView.m (hunk 2 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: Ah okay. Well this PR focused on the resizeMode part, but I had to make some changes to the loading as well to make that work. We should probably use your loading code from the events PR, and then I can integrate the resizeMode changes in there.

File: packages/expo-image/ios/EXImageView.m (hunk 3 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: This was the only way to perform the "post" processing correctly.
I feel that `SDWebImageAvoidAutoSetImage` is pretty much self explanatory

File: packages/expo-image/ios/EXImageView.m (hunk 5 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: No that wouldn't work. The image post-process function only needs to be called whenever the repeat-mode changes (or when the cap-insets change, but that will be added in a later PR)

File: packages/expo-image/ios/EXImageView.m (hunk 6 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: Totally :) I felt this was out-of-scope for this PR, that's why I didn't :)

File: packages/expo-image/ios/EXImageView.m (hunk 7 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: I copied this from RN, but I do agree a helper function would be better. Will do 👍

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java (hunk 10 above)
Comment (IjzerenHein) [reply packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java]: It doesn't. Local assets have a so-called scale-suffix in the file-name (`@3x`) making it possible to who that image by default in it's "real" resolution. This is not possible for remotely loaded assets. For those kinds of images you will have to specify the `width` and `height` in the style or in the `source`.

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java (hunk 9 above)
Comment (IjzerenHein) [reply packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageManager.java]: That is correct. There will be a lot more differences between the iOS/Android implementations at the view/drawing level

File: packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageTypes.java (hunk 8 above)
Comment (IjzerenHein) [reply packages/expo-image/android/src/main/java/expo/modules/image/ExpoImageTypes.java]: Hey! I like this approach and the effort in standardizing this kind of thing.  I've added a dedicated `ExpoImageResizeMode` enum, please let me know whether this is what you had in mind.

While making the change, I came across this bug:
- There seems to be an error in the AudioContentType example that you provided. This line should use `getNativeValue`: https://github.com/expo/expo/blob/d09fc44c62ae96e0347e1909a774b1b93d866b49/packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/enums/AudioContentType.java#L39

File: packages/expo-image/src/ExpoImage.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (sjchmiela) [reply packages/expo-image/src/ExpoImage.tsx]: Would there be any difference if we didn't provide the `width` and `height`? I'm afraid polyfilling this makes a difference in how different images types are handled and rendered which is something I think we want to avoid. 🤔

File: packages/expo-image/ios/EXImageView.m (hunk 5 above)
Comment (sjchmiela) [reply packages/expo-image/ios/EXImageView.m]: True, let me rephrase my concern:

If we set the `contentMode` immediately when the prop is set, on a slow device in the following scenario we would see a glitch:

- an image `X` is downloaded and presented with `resizeMode = A`
- props change: image changes to `Y`, `resizeMode` changes to `B`
  - `setSource:` is called, source updates, `needsReload` is set to `true`
  - `setResizeMode:` is called, `needsReload` is set to `true`, **`contentMode` is changed immediately**
  - `didSetProps:` is called, a new image is downloaded and presented

For a brief moment image `X` is presented with `contentMode` `B`.

File: packages/expo-image/src/ExpoImage.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (IjzerenHein) [reply packages/expo-image/src/ExpoImage.tsx]: I will take a closer look! 👓

File: packages/expo-image/ios/EXImageView.m (hunk 5 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: I understand your concern here and your point is entirely valid.

However, the issue of changing props and then consistently updating the view in one pass (synchronously) would still remain for other props like borders, shadows, etc.. It would just be solved for this particular case..

If you would want to solve this right, you would have to sync all the prop updates on the view (e.g. after the image load has completed). However, this would also affect animatable props and probably make them much update much slower. Personally, I don't think this is a route we would want to go down into.

File: packages/expo-image/src/ExpoImage.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (IjzerenHein) [reply packages/expo-image/src/ExpoImage.tsx]: Okay. So it turns out my suspicion was right here. The `{width, height}` styles are passed down to the Yoga layout system so it can use them as the `intrinsic size` of the image. This means that when you omit any size styles, then the View will use the size of the image itself. 

I've takes a look at doing this at the native level, and it can be done. However, because of the way RN implements this, it has to schedule a block on the UI-manager (which may cause it to wait), and it always causes the layout to be invalidated, causing unnecessary layout passes.
Doing this in JS is really the best performing way as the size information is passed to Yoga directly. I've added comments to the code elaborating on this.

File: packages/expo-image/ios/EXImageView.m (hunk 5 above)
Comment (sjchmiela) [reply packages/expo-image/ios/EXImageView.m]: Sorry for not making myself clear, what I wanted to propose was to _solve things right_ (as you named it) and sync all the prop updates. 😅 When it comes to animatable properties I don't see why would the implementation be slower for them (the only difference that I see would be that in `didSetProps` we would have to compare more values: previous vs next border radii, shadow, etc, which shouldn't be that much slower 🤔).

Looking into other image components (`RCTImageView` and `FFImageView`) I can see they both implement `setResizeMode:` the same way (set the value immediately), so maybe we could get away with it… but I certainly wish we could make sure the scenario above wouldn't happen.

Would you care to add a comment like `// TODO (@sjchmiela): Research props syncing <link to discussion or issue #7440>` so it reminds me to get back to this issue when we're done with major features?

File: packages/expo-image/src/ExpoImage.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (sjchmiela) [reply packages/expo-image/src/ExpoImage.tsx]: I see… my personal opinion on that would be to treat all images alike, so either no user-set size = no image (intrinsic size `(0, 0)`) or to always use image's size as intrinsic size; I think this would be an improvement over current RN's API.

Let's discuss it (its feasibility) in the API proposal (soon to come)! 😃 

Until we decide otherwise I agree with you on the `{width, height}` passed to Yoga. 👍

[the remark this answers is not in the document]
Comment (IjzerenHein) [reply packages/expo-image/src/ExpoImage.tsx]: Let's do that 👍

**Comment by IjzerenHein** [comment]
@sjchmiela I've rebased on the latest master/events. Let me know whether there are any blockers 👍

**Comment by IjzerenHein** [comment]
@sjchmiela Mind if I merge ?

File: packages/expo-image/android/src/main/java/expo/modules/image/enums/ImageResizeMode.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+  
+    public static ImageResizeMode fromStringValue(String value) {
+      for (ImageResizeMode resizeMode : ImageResizeMode.values()) {
+        if (resizeMode.getStringValue().equals(value)) {
+          return resizeMode;
+        }
+      }
+      return ImageResizeMode.UNKNOWN;
+    }
+  }
```
Comment (sjchmiela) [inline packages/expo-image/android/src/main/java/expo/modules/image/enums/ImageResizeMode.java]: This file is unformatted (or formatted wrongly), it has a 4-space indentation (instead of 2-space) and doesn't have a newline at the end of file. I think the AS settings may be git-tracked only for root Android project. 😕

Would you mind reformatting this file? 🙂

File: packages/expo-image/ios/EXImageView.m [hunk 15]
```diff
@@ -2,16 +2,39 @@
...
+@property (nonatomic, strong) NSDictionary *source;
+@property (nonatomic, assign) RCTResizeMode resizeMode;
+@property (nonatomic, assign) BOOL needsReload;
+@property (nonatomic, assign) CGSize intrinsicContentSize;
 
 @end
 
 @implementation EXImageView
+{
+  __weak RCTBridge *_bridge;
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]: Is there any reason for having two types of property declarations (one being with `@property`, other one with `{ properties… }`)? Does this one have any advantage over the other?

If not let's make the property declaration consistent with others:

```objc
@property (nonatomic, weak) RCTBridge *bridge;
```

File: packages/expo-image/ios/EXImageView.m (hunk 7 above)
Comment (sjchmiela) [reply packages/expo-image/ios/EXImageView.m]: Maybe we could:
- either use a static value (as we do with `RCTResizeModeCover`) and put `self.contentMode = UIViewContentModeScaleAspectFill` right after the `_resizeMode` (and maybe write a comment that these two should be kept in sync)
- or use `[self setResizeMode:RCTResizeModeCover]` so that we don't have an initializing code that is somehow special?

File: packages/expo-image/ios/EXImageView.m [hunk 16]
```diff
@@ -43,9 +85,33 @@ - (void)updateImage
     self.onLoadStart(@{});
   }
 
-  [self sd_setImageWithURL:_imageURL
+  NSURL *imageURL = _source ? [RCTConvert NSURL:_source[sourceUriKey]] : nil;
+  NSNumber *scale = _source && _source[sourceScaleKey] ? _source[sourceScaleKey] : nil;
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]:
```suggestion
  NSNumber *scale = _source && _source[sourceScaleKey] ?: nil;
```
Could be written shorter with the Elvis operator

File: packages/expo-image/ios/EXImageView.m [hunk 17]
```diff
@@ -79,6 +149,24 @@ - (SDExternalCompletionBlock)completionBlock
...
+    if (image) {
+      if (resizeMode == RCTResizeModeRepeat) {
+        image = [image resizableImageWithCapInsets:UIEdgeInsetsZero resizingMode:UIImageResizingModeTile];
+      }
+    }
+    
+    // When no explicit source image size was specified, use the dimensions
+    // of the loaded image as the intrinsic content size.
+    if (!width && !height) {
+      [self updateIntrinsicContentSize:image.size internalAsset:NO];
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]:
```suggestion
      [strongSelf updateIntrinsicContentSize:image.size internalAsset:NO];
```

File: packages/expo-image/ios/EXImageView.m [hunk 18]
```diff
@@ -23,12 +46,31 @@ - (void)dealloc
...
+  // so the repeated image fills the UIImageView.
+  self.contentMode = resizeMode == RCTResizeModeRepeat
+    ? UIViewContentModeScaleToFill
+    : (UIViewContentMode)resizeMode;
 }
 
 - (void)didSetProps:(NSArray<NSString *> *)changedProps
 {
-  if ([changedProps containsObject:@"source"]) {
+  if (_needsReload) {
```
Comment (sjchmiela) [inline packages/expo-image/ios/EXImageView.m]: We never set this to `NO` (apart from the initializer).

File: packages/expo-image/src/Image.tsx [hunk 19]
```diff
@@ -1,21 +1,26 @@
...
 
+const DEFAULT_RESIZE_MODE = 'cover';
+
 export interface ImageProps extends AccessibilityProps {
   // On one hand we want to pass resolved source to native module.
   // On the other hand, react-native-web doesn't expose a resolveAssetSource
   // function, so we can't use it there. So we pass the unresolved source
   // to "native components" and they decide whether to resolve the value
   // or not.
+  resizeMode?: ImageResizeMode;
```
Comment (sjchmiela) [inline packages/expo-image/src/Image.tsx]: Could you please put it not immediately after the comment describing the `source` prop (so either before the comment or after the `source` prop) so the comment stays colocated with the prop it talks about?

**Review by sjchmiela — APPROVED** [review]
Apart from these couple of comments it looks good! 👍 👍 👍

File: packages/expo-image/ios/EXImageView.m (hunk 7 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: I've added a helper function which does the conversion. In fact I've added a `EXImageTypes` class which now contains the other enums and helper functions as well, so that the EXImageView class stays a bit cleaner 👍

File: packages/expo-image/android/src/main/java/expo/modules/image/enums/ImageResizeMode.java (hunk 14 above)
Comment (IjzerenHein) [reply packages/expo-image/android/src/main/java/expo/modules/image/enums/ImageResizeMode.java]: Whoops 😅 Reformatted it, thanks! 👍

File: packages/expo-image/ios/EXImageView.m (hunk 15 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: Yeah you're right that would be better. I've updated the code, thanks!

File: packages/expo-image/ios/EXImageView.m (hunk 17 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: Whoops, I missed that one. Thanks, fixed it! 👍

File: packages/expo-image/ios/EXImageView.m (hunk 18 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: Yup I noticed that as well. Fixed now!

File: packages/expo-image/src/Image.tsx (hunk 19 above)
Comment (IjzerenHein) [reply packages/expo-image/src/Image.tsx]: Ah I hadn't noticed that. I thought the props were in alphabetical order that's why I put it there.
I've moved it to the bottom for now.
Would you reckon having the props in alphabetical order is a good thing?

File: packages/expo-image/ios/EXImageView.m (hunk 16 above)
Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: Actually, your suggestion doesn't compile. I was able to write it shorter into

`NSNumber *scale = _source ? _source[sourceScaleKey] : nil;`

Comment (sjchmiela) [reply packages/expo-image/ios/EXImageView.m]: Ah, ok, sorry, didn't notice that the fallback value is `nil`. If so, I think we could just use

```objc
NSNumber *scale = _source[sourceScaleKey];
```

I ensured this should be ok by running

```objc
  NSDictionary *dict = nil;
  NSLog(@"%@", dict[@"abc"]);
```
which didn't crash the application (logged `(null)`).

Comment (IjzerenHein) [reply packages/expo-image/ios/EXImageView.m]: Works like a charm. Must be how Objective-C sends messages to objects that makes this work probably.

Updated the code, looks a lot slimmer 👍
